src/api/ApiResponder.php

```diff
@@ -783,6 +783,121 @@ protected function get_interface_response_dismissGame($interface, $args) {
         return $retval;
     }
 
+    ////////////////////////////////////////////////////////////
+    // Tournament-related methods
+
+    /**
+     * Interface redirect for loadTournaments
+     *
+     * @param BMInterface $interface
+     * @param array $args
+     * @return NULL|array
+     */
+    protected function get_interface_response_loadTournaments($interface) {
+        return $interface->get_all_tournaments($this->session_user_id());
+    }
+
+    /**
+     * Interface redirect for createTournament
+     *
+     * @param BMInterface $interface
+     * @param array $args
+     * @return NULL|array
+     */
+    protected function get_interface_response_createTournament($interface, $args) {
+        return $interface->tournament()->create_tournament(
+            $this->session_user_id(),
+            $args['tournamentType'],
+            $args['nPlayer'],
+            $args['maxWins'],
+            $args['description']
+        );
+    }
+
+    /**
+     * Interface redirect for loadTournamentData
+     *
+     * @param BMInterface $interface
+     * @param array $args
+     * @return NULL|array
+     */
+    protected function get_interface_response_loadTournamentData($interface, $args) {
+        return $interface->load_api_tournament_data(
+            $this->session_user_id(),
+            $args['tournament']
+        );
+    }
+
+    /**
+     * Interface redirect for updateTournament
+     *
+     * @param BMInterface $interface
+     * @param array $args
+     * @return NULL|array
+     */
+    protected function get_interface_response_updateTournament($interface, $args) {
+        if (array_key_exists('buttonNames', $args)) {
+            $button_name_array = json_decode($args['buttonNames'], TRUE);
```

Comment:
This isn't the way to do this --- `ApiSpec` is responsible for decoding types and values --- by the time things get to ApiResponder, they should be their real types.  If we're decoding JSON here, anything could be in it, and that breaks everything we spent a lot of time in #712 trying to make true.

Backing off from philosophy: you can do this the same way `customRecipeArray` is passed to `createGame`, with the addition that we have a special type "button" for a valid API name of a button.  So what you put in ApiSpec is:

```
           'permitted' => array(
                'buttonNames' => array(
                    'arg_type' => 'array',
                    'has_keys' => FALSE,
                    'elem_type' => 'button',
               ),
           ),
```

(I believe that's exactly right, but i didn't test it), and then you pass an array to the API the same way we do for other array parameters like `customRecipeArray`.

Reply:
JS and API spec fixed